pkg/recommendedversion/recommended_version.go

```diff
@@ -0,0 +1,263 @@
+	recommendedArray := strings.Split(recommendedVersionStr, ",")
+
+	// Trim any spaces around the version strings and remove duplicates
+	recommendedVersions := make([]string, 0, len(recommendedArray))
+	alreadyPresent := make(map[string]bool)
+	for _, newVersion := range recommendedArray {
+		trimmedVersion := strings.TrimSpace(newVersion)
+		if trimmedVersion != "" && !alreadyPresent[trimmedVersion] {
+			recommendedVersions = append(recommendedVersions, trimmedVersion)
+			alreadyPresent[trimmedVersion] = true
+		}
+	}
+
+	// Now sort the versions, then reverse the order
+	err := utils.SortVersions(recommendedVersions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reverse the order so it is descending
+	for i := len(recommendedVersions)/2 - 1; i >= 0; i-- {
+		opp := len(recommendedVersions) - 1 - i
+		recommendedVersions[i], recommendedVersions[opp] = recommendedVersions[opp], recommendedVersions[i]
+	}
+	return recommendedVersions, err
+}
+
+func getRecommendationDelayValue() int {
+	delay := recommendedVersionCheckDelaySeconds
+	delayOverride := os.Getenv(constants.ConfigVariableRecommendVersionDelay)
+	if delayOverride != "" {
+		delayOverrideValue, err := strconv.Atoi(delayOverride)
+		if err == nil && delayOverrideValue >= 0 {
+			delay = delayOverrideValue
+		}
+	}
+	return delay
+}
+
+func shouldCheckVersion() bool {
+	delay := getRecommendationDelayValue()
+	if delay == 0 {
+		// The user has disabled the version check
+		return false
+	}
+
+	// Get the last time the version check was done
+	lastCheck := datastore.GetDataStoreValue(dataStoreLastVersionCheckKey)
+	if lastCheck == nil {
+		return true
+	}
+
+	lastCheckTime, ok := lastCheck.(time.Time)
+	if !ok {
+		return true
+	}
+
+	return time.Since(lastCheckTime) > time.Duration(delay)*time.Second
+}
+
+func printVersionRecommendations(writer io.Writer, currentVersion, major, minor, patch string) {
+	if major == "" && minor == "" && patch == "" {
+		// The current version is the best recommended version
+		return
+	}
+
+	// Put a delimiter before this notification so the user
+	// can see it is not part of the command output
+	fmt.Fprintln(writer, "\n==")
+
+	if utils.IsNewVersion(currentVersion, major) || utils.IsNewVersion(currentVersion, minor) || utils.IsNewVersion(currentVersion, patch) {
+		fmt.Fprintf(writer, "WARNING: Due to a problem it is recommended not to use the current version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "Please use a recommended version:")
+	} else {
+		fmt.Fprintf(writer, "Note: A new version of the Tanzu CLI is available. You are at version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "To benefit from the latest security and features, please update to a recommended version:")
+	}
+
+	if major != "" {
+		if utils.IsNewVersion(major, currentVersion) {
+			fmt.Fprintf(writer, "  - %s (new features but breaking changes)\n", major)
+		} else {
+			fmt.Fprintf(writer, "  - %s ([!] you should downgrade to a previous major version)\n", major)
```

Comment:
If I understand this correctly and if we are showing this when the recommended version is lower than currently installed, I am a bit concerned with this warning about downgrading in general. 

1. Why would we want to suggest something like this?
2. What if we released a newer version of the CLI but the central configuration is still recommending an older version? This can easily happen in air-gapped scenarios.

_(Same applies to minor and patch version downgrades as well)_

Reply:
Those are excellent point.  Maybe we should remove the "downgrade" aspect from the feature.

My idea was that let's say we release `v1.3.0` and a couple days after we find it has a serious bug but it will take a couple more days before we can fix it in `v1.3.1`, then we could just remove `v1.3.0` from the list of recommended versions and the user would be told not to use that version.

But your points are too important.  So, if we want to support a "downgrade" aspect, we may need to design more for it.  For example, we could have a different entry in the central config that lists "version not to be used", and only if we add a version that that second list would the WARNING be shown.

But for now, I think I should remove the WARNING.

Thanks catching this problem!

Reply:
I share similar question about major version downgrade message.

I am not sure how the "we do publish plugins earlier" in point 2 is relevant here.

Finally, under what situation can this possibly happen? If the airgap situation is the only realistic situation maybe a different message is warranted ( to the effect of "your airgap central repo state is wrong, contact your administrator if you want to use this version, ...)

For non airgap situations is a scenario where the new CLI major version is picked up but the cached central config state still needs up to default 30m to show up on the CLI host. How do we plan to address this? D we have a means to invalidate the cached data on a new CLI install ?

Reply:
> I am not sure how the "we do publish plugins earlier" in point 2 is relevant here, however. @anujc25

Sorry for the confusion. I mean publish CLI earlier. What I meant is Core CLI gets published to package managers much earlier (1 week) before we announce GA.

Reply:
> My idea was that let's say we release `v1.3.0` and a couple days after we find it has a serious bug but it will take a couple more days before we can fix it in `v1.3.1`, then we could immediately remove `v1.3.0` from the list of recommended versions and the user would be told not to use that version.

Let's consider this another feature that would need more design if we ever choose to add it (I don't think we will 😉 )